### PR TITLE
Issue/pool reporting

### LINKED ISSUE #8228: Update pool metrics collector to also take into account the pool of the schedulers
We collect metrics on pool exhaustion and pool usage. 

We should make sure to add the pools of the schedulers as well 

Notify the FE as well and check the dashboard

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Update database statistics reporting 
1. towards the logs
2. towards the status enpoint
3. towards influxdb

closes #8228 

# TODO post merge:
1. [x]  inform FE that we broke the API
2. [ ] update grafana dashboard on grafana hub

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: mypy-baseline.txt [hunk 1]
```diff
@@ -101,7 +101,7 @@ src/inmanta/ast/type.py:0: error: No overload variant of "reversed" matches argu
 src/inmanta/ast/type.py:0: note: Possible overload variants:
 src/inmanta/ast/type.py:0: note: def [_T] __new__(cls, Reversible[_T], /) -> reversed[_T]
 src/inmanta/ast/type.py:0: note: def [_T] __new__(cls, SupportsLenAndGetItem[_T], /) -> reversed[_T]
-src/inmanta/server/services/databaseservice.py:0: error: Incompatible return value type (got "dict[str, object]", expected "dict[str, BaseModel | UUID | bool | int | float | datetime | str | Sequence[BaseModel | UUID | bool | int | float | datetime | str] | Mapping[str, BaseModel | UUID | bool | int | float | datetime | str]]")  [return-value]
+src/inmanta/server/services/metricservice.py:0: error: Incompatible types in assignment (expression has type "float", variable has type "int")  [assignment]
```
Comment (wouterdb) [inline mypy-baseline.txt:104]: pre-existing issue, moved

File: mypy-baseline.txt [hunk 2]
```diff
@@ -1101,6 +1103,7 @@ src/inmanta/server/services/orchestrationservice.py:0: error: Incompatible types
 src/inmanta/server/services/orchestrationservice.py:0: error: Incompatible types in assignment (expression has type "set[str | None]", variable has type "set[str]")  [assignment]
 src/inmanta/server/services/orchestrationservice.py:0: error: Argument "agents" to "_trigger_auto_deploy" of "OrchestrationService" has incompatible type "Collection[str]"; expected "Sequence[str] | None"  [arg-type]
 src/inmanta/server/services/orchestrationservice.py:0: error: Incompatible return value type (got "ReturnValueWithMeta[Sequence[DesiredStateVersion]]", expected "ReturnValue[list[DesiredStateVersion]]")  [return-value]
+src/inmanta/server/agentmanager.py:0: error: Value of type "dict[str, Any] | None" is not indexable  [index]
```
Comment (wouterdb) [inline mypy-baseline.txt:1106]: decoding of api results, difficult to avoid

File: src/inmanta/agent/agent_new.py [hunk 3]
```diff
@@ -84,6 +86,20 @@ def __init__(
...
+        # Make mypy happy
+        assert data.Resource._connection_pool is not None
+        self._db_monitor = DatabaseMonitor(
+            data.Resource._connection_pool,
+            opt.db_name.get(),
+            opt.db_host.get(),
+            "scheduler",
+            str(self.environment),
+        )
+        self._db_monitor.start()
```
Comment (wouterdb) [inline src/inmanta/agent/agent_new.py:99]: Scheduler gets DB monitor as well

File: src/inmanta/agent/agent_new.py [hunk 4]
```diff
@@ -323,6 +339,20 @@ async def get_facts(self, env: uuid.UUID, agent: str, resource: dict[str, Any])
...
+            return DataBaseReport(
+                connected=False,
+                database="",
+                host="",
+                max_pool=0,
+                open_connections=0,
+                free_connections=0,
+                pool_exhaustion_count=0,
+            )
+        return await self._db_monitor.get_status()
```
Comment (wouterdb) [inline src/inmanta/agent/agent_new.py:354]: new status endpoint on the scheduler to get database status

File: src/inmanta/app.py [hunk 5]
```diff
@@ -156,6 +157,9 @@ async def start() -> None:
             connection_pool_max_size=agent_config.scheduler_db_connection_pool_max_size.get(),
             connection_timeout=agent_config.scheduler_db_connection_timeout.get(),
         )
+        # also report metrics if this is relevant
+        metrics_reporter = MetricsService()
+        metrics_reporter.start_metric_reporters()
```
Comment (wouterdb) [inline src/inmanta/app.py:162]: scheduler will now also report to influxdb

File: src/inmanta/data/model.py [hunk 6]
```diff
@@ -89,7 +89,7 @@ class SliceStatus(BaseModel):
     """
 
     name: str
-    status: dict[str, ArgumentTypes]
+    status: Mapping[str, ArgumentTypes | Mapping[str, ArgumentTypes]]
```
Comment (wouterdb) [inline src/inmanta/data/model.py:92]: This breaks the FE, but allow nicer presentation

File: src/inmanta/server/__init__.py [hunk 7]
```diff
@@ -19,7 +19,7 @@
 # flake8: noqa: F401
 
 SLICE_SERVER = "core.server"
-SLICE_AGENT_MANAGER = "core.agentmanager"
+SLICE_AGENT_MANAGER = "core.scheduler_manager"
```
Comment (wouterdb) [inline src/inmanta/server/__init__.py:22]: rename for FE, hopefully OK

File: src/inmanta/server/agentmanager.py [hunk 8]
```diff
@@ -168,12 +169,63 @@ def __init__(self, closesessionsonstart: bool = True, fact_back_off: Optional[in
 
         self.closesessionsonstart: bool = closesessionsonstart
 
-    async def get_status(self) -> dict[str, ArgumentTypes]:
-        return {
+    async def get_status(self) -> Mapping[str, ArgumentTypes | Mapping[str, ArgumentTypes]]:
```
Comment (wouterdb) [inline src/inmanta/server/agentmanager.py:172]: This method is very tricky, as it has to return withing the timeout of the status endpoint.

I chose to make it report what it has, even if that makes the format somewhat inconsistent. 
(e.g. if the DB is down, the environments will be reported by UUID instead of name)

File: src/inmanta/server/services/databaseservice.py [hunk 9]
```diff
@@ -17,32 +17,144 @@
...
 from inmanta.types import ArgumentTypes
+from inmanta.util import Scheduler
 
 LOGGER = logging.getLogger(__name__)
 
+if typing.TYPE_CHECKING:
+    from inmanta.server import agentmanager
+
+
+class DatabaseMonitor:
```
Comment (wouterdb) [inline src/inmanta/server/services/databaseservice.py:41]: Factored out!

File: src/inmanta/server/services/databaseservice.py [hunk 10]
```diff
@@ -17,32 +17,144 @@
...
+            host=self.db_host,
+            max_pool=self._pool.get_max_size(),
+            open_connections=self._pool.get_size(),
+            free_connections=self._pool.get_idle_size(),
+            pool_exhaustion_count=self._db_pool_watcher._exhausted_pool_events_count,
+        )
+
+    def start_monitor(self) -> None:
+        """Attach to monitoring system"""
+        suffix = f",component={self.component}"
```
Comment (wouterdb) [inline src/inmanta/server/services/databaseservice.py:83]: nasty way of adding tags, but it works

File: tests/test_influxdbreporting.py [hunk 11]
```diff
@@ -182,20 +184,36 @@ async def test_available_metrics(server):
...
-    assert "db.free_connections" in metrics
+    assert metrics["db.connected,component=server"]["value"]
+    assert "db.max_pool,component=server" in metrics
+    assert "db.open_connections,component=server" in metrics
+    assert "db.free_connections,component=server" in metrics
     assert "self.spec.cpu" in metrics
 
     # ensure it doesn't crash when the server is down
     await server.stop()
     metrics = global_registry().dump_metrics()
```
Comment (Hugo-Inmanta) [inline tests/test_influxdbreporting.py:195]:
```suggestion

```

File: stubs/pyformance/__init__.pyi [hunk 12]
```diff
@@ -5,6 +5,9 @@ from pyformance.meters import Gauge
 
 
 class MetricsRegistry:
+
+    _gauges: dict[str, Gauge]
+
```
Comment (Hugo-Inmanta) [inline stubs/pyformance/__init__.pyi:10]: What does this change bring ? I'm not familiar with `.pyi` files

File: src/inmanta/server/agentmanager.py [hunk 13]
```diff
@@ -168,12 +169,66 @@ def __init__(self, closesessionsonstart: bool = True, fact_back_off: Optional[in
...
+        try:
+            # This can timeout, but not likely
+            env_mapping_result = await env_mapping
+            uuid_to_name = {env.id: env.name for env in env_mapping_result}
+        except TimeoutError:
+            # default to uuid's
+            uuid_to_name = {}
+
+        # Filter out timeouts
+        total: DataBaseReport = reduce(lambda x, y: x + y, (x[1] for x in results if not isinstance(x, BaseException)))
```
Comment (Hugo-Inmanta) [inline src/inmanta/server/agentmanager.py:205]:
```suggestion
        total: DataBaseReport | None = reduce(lambda x, y: x + y, (x[1] for x in results if not isinstance(x, BaseException)))
```

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: src/inmanta/data/model.py [hunk 14]
```diff
@@ -859,3 +859,31 @@ def has_source(self) -> bool:
...
+
+    def __add__(self, other: "DataBaseReport") -> "DataBaseReport":
+        if not isinstance(other, DataBaseReport):
+            return NotImplemented
+        if other.database != self.database:
+            return NotImplemented
+        if other.host != self.host:
+            return NotImplemented
+        return DataBaseReport(
+            connected=self.connected,
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/model.py:882]: good question, I should probably `and` them

File: stubs/pyformance/__init__.pyi (hunk 12 above)
Comment (wouterdb) [reply stubs/pyformance/__init__.pyi:10]: it is typing information: pyformance has no mypy type annotation, in this file, we add them.
So I made this field known to mypy by adding it here.

File: src/inmanta/server/agentmanager.py (hunk 13 above)
Comment (wouterdb) [reply src/inmanta/server/agentmanager.py:205]:
```suggestion
        total: DataBaseReport | None = reduce(lambda x, y: x + y, (x[1] for x in results if not isinstance(x, BaseException)), None)
```

Comment (wouterdb) [reply src/inmanta/server/agentmanager.py:205]: good point, but not entirely correct either: 

```
>>> reduce(lambda x,y: x+y, [])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: reduce() of empty iterable with no initial value

```

File: src/inmanta/server/agentmanager.py [hunk 15]
```diff
@@ -251,6 +252,12 @@ async def start(self) -> None:
             await self._expire_all_sessions_in_db()
 
         self.add_background_task(self._process_session_listener_actions())
+        # Schedule cleanup agentprocess and agentinstance tables
+        agent_process_purge_interval = opt.agent_process_purge_interval.get()
+        if agent_process_purge_interval > 0:
+            self.schedule(
+                self._purge_agent_processes, interval=agent_process_purge_interval, initial_delay=0, cancel_on_stop=False
+            )
```
Comment (Hugo-Inmanta) [inline src/inmanta/server/agentmanager.py:260]: Why was this moved ?

**Review by Hugo-Inmanta — CHANGES_REQUESTED** [review]

File: src/inmanta/server/agentmanager.py (hunk 15 above)
Comment (wouterdb) [reply src/inmanta/server/agentmanager.py:260]: Because I think it should have been in the agent manager in the first place: it is related to management of agents, so it should be in the agent manager.

**Review by Hugo-Inmanta — APPROVED** [review]

File: changelogs/unreleased/8228-database_pool_report.yml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+description: Update database pool status reporting for the new scheduler
+issue-nr: 8228
+change-type: minor
+destination-branches: [master]
+sections:
+  minor-improvement: "{{description}}"
```
Comment (sanderr) [inline changelogs/unreleased/8228-database_pool_report.yml:6]: Does this deserve a change entry, considering that the new scheduler has not been publicly released? Or has it?

File: src/inmanta/server/agentmanager.py [hunk 17]
```diff
@@ -168,12 +169,67 @@ def __init__(self, closesessionsonstart: bool = True, fact_back_off: Optional[in
...
             "sessions": len(self.sessions),
         }
 
+        # Try to get more info from scheduler, but make sure not to timeout
+        schedulers = self.get_all_schedulers()
+        deadline = 0.9 * Server.GET_SERVER_STATUS_TIMEOUT
+
+        async def get_report(env: uuid.UUID, session: protocol.Session) -> tuple[uuid.UUID, DataBaseReport]:
+            result = await asyncio.wait_for(session.client.get_db_status(), deadline)
+            assert result.code == 200
```
Comment (sanderr) [inline src/inmanta/server/agentmanager.py:186]: Is assert appropriate here? It can realistically fail, can't it?

File: src/inmanta/server/agentmanager.py [hunk 18]
```diff
@@ -168,12 +169,67 @@ def __init__(self, closesessionsonstart: bool = True, fact_back_off: Optional[in
...
+            # Mypy can't help here, ....
```
Comment (sanderr) [inline src/inmanta/server/agentmanager.py:187]: Well,  `assert result.result is not None` would probably do it. Still far from perfect.

File: src/inmanta/util/__init__.py [hunk 19]
```diff
@@ -392,6 +392,25 @@ def action_function() -> None:
         self._scheduled[task_spec] = handle
         return task_spec
 
+    def schedule(
```
Comment (sanderr) [inline src/inmanta/util/__init__.py:395]: What does this method offer? I do prefer the name, but I feel like it's just a layer on top of `add_action` that makes a tiny streamline of the interface for one type of schedules, at the cost of dropping support for the other. So now we need to call either `schedule()` or `add_action()` depending on the use case.

File: src/inmanta/server/protocol.py [hunk 20]
```diff
@@ -399,7 +390,7 @@ def get_extension_statuses(cls, slices: list["ServerSlice"]) -> list[ExtensionSt
                 result[ext_status.name] = ext_status
         return list(result.values())
 
-    async def get_status(self) -> dict[str, ArgumentTypes]:
+    async def get_status(self) -> Mapping[str, ArgumentTypes | Mapping[str, ArgumentTypes]]:
```
Comment (sanderr) [inline src/inmanta/server/protocol.py:393]: Why change all these from `dict` to `Mapping`?

**Review by sanderr — CHANGES_REQUESTED** [review]
I'm only vaguelly familiar with all of this tbh.

File: changelogs/unreleased/8228-database_pool_report.yml (hunk 16 above)
Comment (wouterdb) [reply changelogs/unreleased/8228-database_pool_report.yml:6]: It is a user visible change, but perhaps I should add more detail?

File: src/inmanta/server/agentmanager.py (hunk 18 above)
Comment (wouterdb) [reply src/inmanta/server/agentmanager.py:187]: problem is that you also has to assert the result is a dict, containing `data` which is also a dict, ....

File: src/inmanta/server/agentmanager.py (hunk 17 above)
Comment (wouterdb) [reply src/inmanta/server/agentmanager.py:186]: Good question, it is a bit a special case as the exception is handled right next to it and logged as debug, so it matters little, but I can improve if desired. 

This code is very very robust, making it less failure transparent.

File: src/inmanta/server/protocol.py (hunk 20 above)
Comment (wouterdb) [reply src/inmanta/server/protocol.py:393]: Due to invariance of dict, the Union is a bit hard to use otherwise  and it is inteded to be mapping all along

File: src/inmanta/util/__init__.py (hunk 19 above)
Comment (sanderr) [reply src/inmanta/util/__init__.py:395]: To me it seems like it adds more confusion than anything else. If I see a method `schedule` I'd assume that to be the main interface, but it isn't because it only supports one of the two schedule types.

So I prefer it without, but if you feel it's useful I won't stand in the way.

File: changelogs/unreleased/8228-database_pool_report.yml (hunk 16 above)
Comment (sanderr) [reply changelogs/unreleased/8228-database_pool_report.yml:6]: I mean the entire scheduler is new, so I would assume that to be one change entry rather than a separate one for each prt it affects.

File: src/inmanta/server/agentmanager.py (hunk 18 above)
Comment (sanderr) [reply src/inmanta/server/agentmanager.py:187]: True

File: src/inmanta/server/protocol.py (hunk 20 above)
Comment (wouterdb) [reply src/inmanta/server/protocol.py:393]: Left it like this, I think Mapping better reflects the intent.

File: changelogs/unreleased/8228-database_pool_report.yml (hunk 16 above)
Comment (wouterdb) [reply changelogs/unreleased/8228-database_pool_report.yml:6]: I don't want to lose the details, but we can probably collect them at the end

File: src/inmanta/server/protocol.py (hunk 20 above)
Comment (sanderr) [reply src/inmanta/server/protocol.py:393]: I don't necessarily agree as a general rule, but I can see your point for this case.

File: mypy-baseline.txt [hunk 21]
```diff
@@ -740,6 +739,7 @@ src/inmanta/ast/variables.py:0: error: Missing type parameters for generic type
 src/inmanta/ast/variables.py:0: error: Missing type parameters for generic type "VariableReader"  [type-arg]
 src/inmanta/ast/variables.py:0: error: Argument 2 to "AttributeReference" has incompatible type "str"; expected "LocatableString"  [arg-type]
 src/inmanta/server/protocol.py:0: error: Argument 1 to "RESTServer" has incompatible type "SessionManager"; expected "SessionManagerInterface"  [arg-type]
+src/inmanta/server/protocol.py:0: error: Name "IntervalSchedule" is not defined  [name-defined]
```
Comment (sanderr) [inline mypy-baseline.txt:742]: This looks like a missing import to me.

**Review by sanderr — APPROVED** [review]
One comment but I expect the tests will fail on it so I'm approving.

File: mypy-baseline.txt (hunk 21 above)
Comment (wouterdb) [reply mypy-baseline.txt:742]: it is
